test/NameBinding/dynamic-member-lookup.swift

```diff
@@ -0,0 +1,231 @@
+  subscript(dynamicMember member: String...) -> Int {
+    return 42
+  }
+}
+
+// References to overloads are resolved just like normal subscript lookup:
+// they are either contextually disambiguated or are invalid.
+@dynamicMemberLookup
+struct Ambiguity {
+  subscript(dynamicMember member: String) -> Int {
+    return 42
+  }
+  subscript(dynamicMember member: String) -> Float {
+    return 42
+  }
+}
+
+func testAmbiguity(a : Ambiguity) {
+  let _ : Int = a.flexibility
+  let _ : Float = a.dynamism
+  _ = a.dynamism  // expected-error {{ambiguous use of 'subscript(dynamicMember:)'}}
+}
+
+
+// expected-error @+1 {{'@dynamicMemberLookup' attribute cannot be applied to this declaration}}
+@dynamicMemberLookup
+extension Int {
+  subscript(dynamicMember member: String) -> Int {
+    fatalError()
+  }
+}
+
+@dynamicMemberLookup  // expected-error {{'@dynamicMemberLookup' attribute cannot be applied to this declaration}}
+func NotAllowedOnFunc() {
+}
+
+//===----------------------------------------------------------------------===//
+// Test Existential
+//===----------------------------------------------------------------------===//
+
+@dynamicMemberLookup
+protocol PyVal {
+  subscript(dynamicMember member: StaticString) -> PyVal { get nonmutating set }
+}
+extension PyVal {
+  subscript(dynamicMember member: StaticString) -> PyVal {
+    get { fatalError() } nonmutating set {}
+  }
+}
+
+struct MyType : PyVal {
+}
+
+
+func testMutableExistential(a : PyVal, b : MyType) -> PyVal {
+  a.x.y = b
+  b.x.y = b
+  return a.foo.bar.baz
+}
+
+//===----------------------------------------------------------------------===//
+// JSON example
+//===----------------------------------------------------------------------===//
+
+@dynamicMemberLookup
+enum JSON {
+  case IntValue(Int)
+  case StringValue(String)
+  case ArrayValue(Array<JSON>)
+  case DictionaryValue(Dictionary<String, JSON>)
+
+  var stringValue : String? {
+    if case .StringValue(let str) = self {
+      return str
+    }
+    return nil
+  }
+  subscript(index: Int) -> JSON? {
+    if case .ArrayValue(let arr) = self {
+      return index < arr.count ? arr[index] : nil
+    }
+    return nil
+  }
+  subscript(key: String) -> JSON? {
+    if case .DictionaryValue(let dict) = self {
+      return dict[key]
+    }
+    return nil
+  }
+  
+  subscript(dynamicMember member: String) -> JSON? {
+    if case .DictionaryValue(let dict) = self {
+      return dict[member]
+    }
+    return nil
+  }
+}
+func test_json_example(x : JSON) -> String? {
+  _ = x.name?.first
+  return x.name?.first?.stringValue
+}
+
+//===----------------------------------------------------------------------===//
+// Derived Class Example
+//===----------------------------------------------------------------------===//
+
+@dynamicMemberLookup
+class BaseClass {
+  subscript(dynamicMember member: String) -> Int {
+    return 42
+  }
+}
+class DerivedClass : BaseClass {
+}
```

Comment:
Fun case to try: a derived class overrides `subscript(dynamicMember:)` to make it settable.

Reply:
Works 👍, testcase added.